plugins/package-managers/node/src/main/kotlin/Yarn2.kt

```diff
@@ -157,6 +171,8 @@ class Yarn2(
     override fun command(workingDir: File?): String {
         if (workingDir == null) return ""
 
+        if (options[OPTION_COREPACK_ENABLED].toBoolean()) return DEFAULT_EXECUTABLE_NAME
```

Comment:
IMO it would be nicer if this could be (reliably) auto-detected. Does the `yarn exec env` approach mentioned over [here](https://yarnpkg.com/corepack#installation) work for that?

Reply:
This is an interesting point, I was also thinking about this, but am not sure how to handle this in the best way. It would also be possible to even go a step further and design the `Yarn2` class in a way that it - optionally - executes the `corepack enable` command, which is required to use Corepack.

My line of thinking was that ORT is rather low-level. It should focus on obtaining the dependencies from package managers and assume that the environment is properly set up. The setup of the environment falls in the responsibility of integrators, e.g. ORT Server. (This is also related to the discussion we had in the last ORT Community meeting.) Now, the integration environment typically knows whether Corepack should be used or not and can directly provide the information to `Yarn2`; so there is no need to do an additional check.

But I am open for other opinions.

Reply:
Any other opinions or comments on that?

Reply:
Coincidentally, I've read about Corepack today also in the context of https://github.com/oss-review-toolkit/ort/issues/8715. Could we simply auto-enable the use of Corepack if `package.json` has the [`packageManager` node](https://nodejs.org/api/packages.html#packagemanager)?

Reply:
Well, this does not really play well with the API: Whether Corepack is enabled or not affects the `command` function of `Yarn2`. However, this function has no obvious way to check the status of Corepack. It is called early in the analysis process to determine the tool version. IIUC, at that time, `Yarn2` does not know about the definition files it will have to process.

Reply:
This relies on `resolveDependencies` always being called before `command()`, which may change in a refactoring.

But aside from that, this brings me back to my original question: Even if feasible, should the responsibility to determine the Corepack status not be shifted to the execution environment of ORT? Here, the information is typically available without major hickhacks.

Reply:
Ah, yes, I forgot to reply to that earlier:

> My line of thinking was that ORT is rather low-level. It should focus on obtaining the dependencies from package managers and assume that the environment is properly set up.

I'd agree to the second sentence, but not so much to the former. Whenever feasible to detect and apply, ORT should use the correct tools and versions automatically. Just out of pure pragmatism we deviated from this rule of thumb for ecosystems where there's no single defined way to declare the tool version, or bootstrapping the tool (on all supported platform) would be infeasible / too much effort.

Or phrased differently: Development teams that take ORT into use should not be required to configure their tools / version for ORT again if they already did so in their build system, as we'd want to avoid teams to maintain basically the same information in multiple different places. That thinking is also the motivation behind the already linked https://github.com/oss-review-toolkit/ort/issues/8715.

However, whether the auto-detection implementation is feasible needs to be decided on a case-by-case basis. And for this particular Yarn / Corepack case, I'm not yet convinced that is is not feasible.

Reply:
Then I propose the following: The default behavior is to auto-detect whether Corepack is enabled based on the content of the `package.json` file. But this behavior can be overridden with the configuration property; then this mechanism can be disabled if it is not needed.

Would this be okay?

Reply:
I have changed the implementation accordingly.

Reply:
Thanks. But I wonder whether we really needs this:

> But this behavior can be overridden with the configuration property

IMO too many things in ORT are configurable already, not always with a proven use-case (anymore). So, you do already see a concrete case where one would opt-out of auto-detection, or was the configuration added "just in case"?

Reply:
Yes, for ORT Server (at least our installation), we will probably enable Corepack by default. So, the result of the auto-detection must be overridden. The reason for this is that Corepack does not support other installation options; for instance, it refuses to work with a version of yarn that was installed via the package manager of the OS. This means in our case that if we want to support Corepack, we cannot provide default installations of certain package managers, but must rely on Corepack to do the installation.